src/http-api/resources/files.js

```diff
@@ -33,18 +40,142 @@ exports.cat = {
   // main route handler which is called after the above `parseArgs`, but only if the args were valid
   handler: (request, reply) => {
     const key = request.pre.args.key
+    const ipfs = request.server.app.ipfs
 
-    request.server.app.ipfs.files.cat(key, (err, stream) => {
+    ipfs.files.cat(key, (err, stream) => {
       if (err) {
         log.error(err)
         return reply({
           Message: 'Failed to cat file: ' + err,
           Code: 0
         }).code(500)
       }
-      stream.on('data', (data) => {
-        return reply(data)
+
+      // hapi is not very clever and throws if no
+      // - _read method
+      // - _readableState object
+      // are there :(
+      stream._read = () => {}
+      stream._readableState = {}
+      return reply(stream).header('X-Stream-Output', '1')
+    })
+  }
+}
+
+exports.get = {
+  // uses common parseKey method that returns a `key`
+  parseArgs: exports.parseKey,
+
+  // main route handler which is called after the above `parseArgs`, but only if the args were valid
+  handler: (request, reply) => {
+    const key = request.pre.args.key
+    const ipfs = request.server.app.ipfs
+    const pack = tar.pack()
+
+    ipfs.files.getPull(key, (err, stream) => {
+      if (err) return handleError(err)
+
+      pull(
+        stream,
+        pull.asyncMap((file, cb) => {
+          const header = {name: file.path}
+
+          if (!file.content) {
+            header.type = 'directory'
+            pack.entry(header)
+            cb()
+          } else {
+            header.size = file.size
+            toStream.source(file.content)
+              .pipe(pack.entry(header, cb))
+          }
+        }),
+        pull.onEnd((err) => {
+          if (err) return handleError(err)
+
+          pack.finalize()
+          reply(pack).header('X-Stream-Output', '1')
+        })
+      )
+    })
+
+    function handleError (err) {
```

Comment:
Might not be a great idea to create this function every time a request happens. And in general, `handleError` should probably be somewhere global for the lib instead of specific to get.